// projekt24/src/app/login/login.component.ts
import { Component, OnInit } from '@angular/core';
import {Router} from '@angular/router';
import { FormGroup, FormBuilder, FormControl, Validators } from '@angular/forms';
import { HttpClient } from '@angular/common/http';
import { AuthenticationService } from '../services/authentication.service';
import { HttpClientService } from '../services/http-client.service';

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.css']
})
export class LoginComponent implements OnInit {

  loginForm: FormGroup;
  errorMsg;

  constructor(private auth:AuthenticationService, private fb:FormBuilder, private router:Router, private http:HttpClient, private httpClient: HttpClientService) { }

  ngOnInit(): void {
    this.loginForm = this.fb.group({
      'username' : new FormControl(null, Validators.required),
      'password' : new FormControl(null, Validators.required)
    });

    this.auth.errorMessage.subscribe(e=>this.errorMsg=e);
    
  }
  
  onSubmit(){
    let credentials=this.loginForm.value
    this.auth.onLogin(credentials);
  }

}
