// projekt24/src/app/register/register.component.ts
import { Component, OnInit } from '@angular/core';
import {FormBuilder, FormControl, FormGroup, Validators} from "@angular/forms";
import {Subject} from "rxjs";
import {Router} from "@angular/router";
import {HttpClient} from '@angular/common/http';
import { AuthenticationService } from '../services/authentication.service';


@Component({
  selector: 'app-register',
  templateUrl: './register.component.html',
  styleUrls: ['./register.component.css']
})
export class RegisterComponent implements OnInit {

  registerForm: FormGroup;

  private user;
  errorMsg: string=null;

  constructor(private fb: FormBuilder , private router: Router , private http: HttpClient , private auth: AuthenticationService) { }

  ngOnInit() {

    this.registerForm=new FormGroup({
      'username' : new FormControl(null, [Validators.required, Validators.minLength(4)]),
      'password' : new FormControl(null, [Validators.required, Validators.minLength(4)]),
      'password2' : new FormControl(null, [Validators.required, Validators.minLength(4)]),
      'name' : new FormControl(null, Validators.required),
      'email':new FormControl(null, [Validators.required, Validators.email]),
      'status':new FormControl(null, Validators.required)
    });

    /*this.auth.errorEmitter.subscribe(err=>this.errorMsg=err);*/

  }

  onSubmit(){
    let credentials=this.registerForm.value;
    if(this.registerForm.get('password').value == this.registerForm.get('password2').value){
      delete credentials.password2;
      this.auth.onRegister(credentials);
      console.log(credentials);
    }
  }

}
