// projekt24/src/app/services/http-client.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import {map} from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class HttpClientService {

  url:String="https://th-8f0a5-default-rtdb.firebaseio.com/";

  constructor(private http:HttpClient) { }

  getData(table){
    return this.http.get(this.url+table+'.json').pipe(map(res=>{
      let data=[];
      for(let elem in res){
        data.push({...res[elem],id:elem});
      }
      return data
    }));
  }

  getUsers(){
    return this.getData('users');
  }
  addUser(user){
    return this.http.post(this.url+'users.json',user);
  }
  deleteUser(id){
    return this.http.delete(this.url+`users/${id}.json`);
  }
  editUser(user){
    return this.http.patch(this.url+`users/${user.id}/.json`,user);
  }
  
  getpPilots(){
    return this.getData('pPilots');
  }
  addpPilot(pPilot){
    return this.http.post(this.url+'pPilots.json',pPilot);
  }
  deletepPilot(id){
    return this.http.delete(this.url+`pPilots/${id}.json`);
  }
  editpPilot(pPilot){
    let id=pPilot.id;
    delete pPilot.id;
    return this.http.patch(this.url+`pPilots/${id}/.json`,pPilot);
  }

  getpMechs(){
    return this.getData('pMechs');
  }
  addpMech(pMech){
    return this.http.post(this.url+'pMechs.json',pMech);
  }
  deletepMech(id){
    return this.http.delete(this.url+`pMechs/${id}.json`);
  }
  editpMech(pMech){
    let id=pMech.id;
    delete pMech.id;
    return this.http.patch(this.url+`pMechs/${id}/.json`,pMech);
  }

  getePilots(){
    return this.getData('ePilots');
  }
  addePilot(ePilot){
    return this.http.post(this.url+'ePilots.json',ePilot);
  }
  deleteePilot(id){
    return this.http.delete(this.url+`ePilots/${id}.json`);
  }
  editePilot(ePilot){
    let id=ePilot.id;
    delete ePilot.id;
    return this.http.patch(this.url+`ePilots/${id}/.json`,ePilot);
  }

  geteMechs(){
    return this.getData('eMechs');
  }
  addeMech(eMech){
    return this.http.post(this.url+'eMechs.json',eMech);
  }
  deleteeMech(id){
    return this.http.delete(this.url+`eMechs/${id}.json`);
  }
  editeMech(eMech){
    let id=eMech.id;
    delete eMech.id;
    return this.http.patch(this.url+`eMechs/${id}/.json`,eMech);
  }

}
