// projekt24/src/app/profile/profile.component.ts
import { Component, OnInit } from '@angular/core';
import {Router} from '@angular/router';
import { AuthenticationService } from '../services/authentication.service';
import { HttpClientService } from '../services/http-client.service';

@Component({
  selector: 'app-profile',
  templateUrl: './profile.component.html',
  styleUrls: ['./profile.component.css']
})
export class ProfileComponent implements OnInit {

  currUser;
  pPilots;
  pMechs;
  ePilots;
  eMechs;

  filterPilotsFlag=false;
  filterMechsFlag=false;

  constructor(private router: Router, private auth: AuthenticationService, private httpClient:HttpClientService) { }

  ngOnInit(): void {
    this.currUser=JSON.parse(localStorage.getItem('user'));

    if(!this.currUser){
      this.router.navigate(['/login']);
    }
    this.httpClient.getpPilots().subscribe(res=>{
      this.pPilots=res;
    });
    this.httpClient.getpMechs().subscribe(res=>{
      this.pMechs=res;
    });
    this.httpClient.getePilots().subscribe(res=>{
      this.ePilots=res;
    });
    this.httpClient.geteMechs().subscribe(res=>{
      this.eMechs=res;
    });
    
  }

  logout(){
    this.auth.onLogout();
  }

  onpPilotDelete(id){
    this.httpClient.deletepPilot(id).subscribe(res=>{
      this.pPilots=this.pPilots.filter(e=>e.id!=id);
    });
  }

  onpMechDelete(id){
    this.httpClient.deletepMech(id).subscribe(res=>{
      this.pMechs=this.pMechs.filter(e=>e.id!=id);
    });
  }

  onePilotDelete(id){
    this.httpClient.deleteePilot(id).subscribe(res=>{
      this.ePilots=this.ePilots.filter(e=>e.id!=id);
    });
  }

  oneMechDelete(id){
    this.httpClient.deleteeMech(id).subscribe(res=>{
      this.eMechs=this.eMechs.filter(e=>e.id!=id);
    });
  }

}
